p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,281 @@
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multibase"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/pion/ice/v2"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ tpt.Listener = &listener{}
+
+const (
+	maxConnectionBufferSize int = 10
+	candidateSetupTimeout       = 20 * time.Second
+	candidateChanSize           = 20
+)
+
+type candidateAddr struct {
+	ufrag string
+	raddr *net.UDPAddr
+}
+
+type listener struct {
+	transport                 *WebRTCTransport
+	config                    webrtc.Configuration
+	localFingerprint          webrtc.DTLSFingerprint
+	localFingerprintMultibase string
+	mux                       ice.UDPMux
+	ctx                       context.Context
+	cancel                    context.CancelFunc
+	localAddr                 net.Addr
+	localMultiaddr            ma.Multiaddr
+	connChan                  chan tpt.CapableConn
+	wg                        sync.WaitGroup
+}
+
+func newListener(transport *WebRTCTransport, laddr ma.Multiaddr, socket net.PacketConn, config webrtc.Configuration) (*listener, error) {
+	candidateChan := make(chan candidateAddr, candidateChanSize)
+	mux := udpmux.NewUDPMux(socket, func(ufrag string, addr net.Addr) {
+		// Push to the candidateChan asynchronously to avoid blocking the mux goroutine
+		// on candidates being processed. This can cause new connections to fail at high
+		// throughput but will allow packets for existing connections to be processed.
+		select {
+		case candidateChan <- candidateAddr{ufrag: ufrag, raddr: addr.(*net.UDPAddr)}:
+		default:
+		}
+	})
+	localFingerprints, err := config.Certificates[0].GetFingerprints()
+	if err != nil {
+		return nil, err
+	}
+
+	localMh, err := hex.DecodeString(strings.ReplaceAll(localFingerprints[0].Value, ":", ""))
+	if err != nil {
+		return nil, err
+	}
+	localMhBuf, _ := multihash.Encode(localMh, multihash.SHA2_256)
+	localFpMultibase, _ := multibase.Encode(multibase.Base58BTC, localMhBuf)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	l := &listener{
+		mux:                       mux,
+		transport:                 transport,
+		config:                    config,
+		localFingerprint:          localFingerprints[0],
+		localFingerprintMultibase: localFpMultibase,
+		localMultiaddr:            laddr,
+		ctx:                       ctx,
+		cancel:                    cancel,
+		localAddr:                 socket.LocalAddr(),
+		connChan:                  make(chan tpt.CapableConn, maxConnectionBufferSize),
+	}
+
+	l.wg.Add(1)
+	go l.handleIncomingCandidates(candidateChan)
+	return l, err
+}
+
+func (l *listener) handleIncomingCandidates(candidateChan chan candidateAddr) {
+	defer l.wg.Done()
+	for {
+		select {
+		case <-l.ctx.Done():
+			return
+		case addr := <-candidateChan:
+			go func() {
+				ctx, cancelFunc := context.WithTimeout(context.Background(), candidateSetupTimeout)
+				defer cancelFunc()
```

Comment:
Nit: more idiomatic naming
```suggestion
				ctx, cancel := context.WithTimeout(context.Background(), candidateSetupTimeout)
				defer cancel()
```